src/app/content/launchToken.ts

```diff
@@ -0,0 +1,44 @@
+import { Window } from '@openstax/types/lib.dom';
+import { useServices } from '../context/Services';
+
+export const decodeToken = (launchToken: string | undefined) => {
+  if (!launchToken || typeof window === 'undefined') return undefined;
+
+  // https://stackoverflow.com/a/38552302/14809536
+  const base64Url = launchToken.split('.')[1];
+  const base64 = base64Url.replace(/-/g, '+').replace(/_/g, '/');
+  const jsonPayload = decodeURIComponent(window.atob(base64).split('').map((c) => {
+      return '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2);
+  }).join(''));
+
+  const token = JSON.parse(jsonPayload);
+  return 'sub' in token ? JSON.parse(token.sub) : undefined;
+};
+
+export const pullToken = (window: Window) => {
+  const searchParams = new URLSearchParams(window.location.search);
+
+  const launchToken = searchParams.get('t') ?? undefined;
```

Comment:
The `??` branch seems pointless.

Reply:
you're right, the conditional below catches it either way